# Homework 2

In [36]:
using LinearAlgebra

#### 8. Assume that you have a two link planar manipulator. $\theta_1$ is the angle between the x axis (measured counter-clockwise as positive) and the first link arm. $\theta_2$ is the angle between the second link arm and the first link arm (again measured counter-clockwise as positive). The link lengths are $a_1$, $a_2$. Derive the formulas for $dx/dt$ and $dy/dt$ as a function of $d\theta_1/dt$ and $d\theta_2/dt$.

$$ x = a_2cos(\theta_1 + \theta_2) + a_1cos(\theta_1) $$

$$ y = a_2sin(\theta_1 + \theta_2) + a_1sin(\theta_1) $$

$$ J_f = \begin{pmatrix} \frac{\partial x}{\partial\theta_1} & \frac{\partial x}{\partial\theta_2} \\ \frac{\partial y}{\partial\theta_1} & \frac{\partial y}{\partial\theta_2} \end{pmatrix}, v = J_f\dot q $$

$$ v = \begin{pmatrix} \frac{dx}{dt} \\ \frac{dy}{dt} \end{pmatrix} = \begin{pmatrix}
-a_2sin(\theta_1 + \theta_2) - a_1sin(\theta_1) & -a_2sin(\theta_1 + \theta_2)\\
a_2cos(\theta_1 + \theta_2) + a_1cos(\theta_1) & a_2cos(\theta_1 + \theta_2)
\end{pmatrix} \begin{pmatrix} \frac{d\theta_1}{dt} \\ \frac{d\theta_2}{dt} \end{pmatrix} $$

#### 9. Using the two link planar manipulator described in the previous problem. Derive the formulas for $d\theta_1/dt$ and $d\theta_2/dt$ as a function of $x$, $y$, $dx/dt$ and $dy/dt$.

$$ D = \frac{x^2 + y^2 - a^2_1 - a^2_2}{2a_1a_2} $$

$$
\theta_2 = tan^{-1}\left(\frac{\pm\sqrt{1 - D^2}}{D}\right), \theta_1 = tan^{-1}\left(\frac{y}{x}\right) - 
tan^{-1}\left(
    \frac{a_2sin\left(
        \theta_2
    \right)}{a_1 + a_2cos\left(
        \theta_2
    \right)}
\right)
$$

$$ J^{-1}_f = \frac{1}{\frac{\partial x}{\partial\theta_1}\frac{\partial y}{\partial\theta_2} - \frac{\partial x}{\partial\theta_2}\frac{\partial y}{\partial\theta_1}} \begin{pmatrix} \frac{\partial x}{\partial\theta_1} & \frac{\partial x}{\partial\theta_2} \\ \frac{\partial y}{\partial\theta_1} & \frac{\partial y}{\partial\theta_2} \end{pmatrix}, \dot q = J^{-1}_fv $$

$$ \dot q = \begin{pmatrix} \frac{d\theta_1}{dt} \\ \frac{d\theta_2}{dt} \end{pmatrix} = \frac{1}{a_1a_2sin(\theta_2)} \begin{pmatrix}
a_2cos(\theta_1 + \theta_2) & a_2sin(\theta_1 + \theta_2)\\
-a_2cos(\theta_1 + \theta_2) - a_1cos(\theta_1) & -a_2sin(\theta_1 + \theta_2) - a_1sin(\theta_1)
\end{pmatrix} \begin{pmatrix} \frac{dx}{dt} \\ \frac{dy}{dt} \end{pmatrix} $$

#### 10. Given the two link manipulator as described in the previous problems with the length of the first link $a_1 = 10$ and the second link $a_2 = 10$.

##### a. If $\theta_1 = 45^{\circ}$, $\theta_2 = 45^{\circ}$, find $x$ and $y$.
$$ x = 10cos(90^{\circ}) + 10cos(45^{\circ}) = 5\sqrt{2} $$

$$ y = 10sin(90^{\circ}) + 10sin(45^{\circ}) = 10 + 5\sqrt{2} $$

##### b. If $\theta_1 = 45^{\circ}$, $\theta_2 = 45^{\circ}$, $d\theta_1/dt = 5^{\circ}s^{-1}$, $d\theta_2/dt = 10^{\circ}s^{-1}$ find $dx/dt$ and $dy/dt$.
$$ \theta_1 = \frac{\pi}{4}, \theta_2 = \frac{\pi}{4}, \frac{d\theta_1}{dt} = \frac{\pi}{36}s^{-1}, \frac{d\theta_2}{dt} = \frac{\pi}{18}s^{-1} $$

$$ \begin{pmatrix} \frac{dx}{dt} \\ \frac{dy}{dt} \end{pmatrix} = \begin{pmatrix}
-10sin(\frac{\pi}{2}) - 10sin(\frac{\pi}{4}) & -10sin(\frac{\pi}{2})\\
10cos(\frac{\pi}{2}) + 10cos(\frac{\pi}{4}) & 10cos(\frac{\pi}{2})
\end{pmatrix} \begin{pmatrix} \frac{\pi}{36}s^{-1} \\ \frac{\pi}{18}s^{-1} \end{pmatrix} $$

$$ \frac{dx}{dt} = \frac{-5\pi}{18}\left(3 + \frac{\sqrt{2}}{2}\right)s^{-1} $$

$$ \frac{dy}{dt} = \frac{5\pi\sqrt{2}}{36}s^{-1} $$

#### 11. Given the two link manipulator as above with the length of the first link $a_1 = 10$ and the second link $a_2 = 10$.

##### a. If $x = 10$, $y = 14$, find $\theta_1$ and $\theta_2$.
$$ D = \frac{10^2 + 14^2 - 10^2 - 10^2}{200} = 0.48 $$

In [37]:
t2 = atan(sqrt(1 - 0.48*0.48), 0.48)
println(t2, " rads  |  ", t2*180/pi, " degs")

1.0701416143903084 rads  |  61.31459798588107 degs


$$ \theta_2 = tan^{-1}\left(\frac{\pm\sqrt{1 - 0.48^2}}{0.48}\right) = 61.315^{\circ} $$

In [38]:
t1 = atan(14, 10) - atan(10*sin(t2), 10 + 10*cos(t2))
println(t1, " rads  |  ", t1*180/pi, " degs")

0.41547603361692076 rads  |  23.80502321508507 degs


$$\theta_1 = tan^{-1}\left(\frac{14}{10}\right) - tan^{-1}\left(\frac{10sin(1.070)}{10 + 10cos(1.070)}\right) = 23.805^{\circ} $$

##### b. If $x = 10$, $y = 14$, $dx/dt = -0.25s^{-1}$, $dy/dt = 0.5s^{-1}$, find $d\theta_1/dt$ and $d\theta_2/dt$.
$$ \begin{pmatrix} \frac{d\theta_1}{dt} \\ \frac{d\theta_2}{dt} \end{pmatrix} = \frac{1}{100sin(1.070)} \begin{pmatrix}
10cos(0.415 + 1.070) & 10sin(0.415 + 1.070) \\
-10cos(0.415 + 1.070) - 10cos(0.415) & -10sin(0.415 + 1.070) - 10sin(0.415)
\end{pmatrix} \begin{pmatrix} -0.25s^{-1} \\ 0.5s^{-1} \end{pmatrix} $$

In [39]:
det = 100*sin(t2)
JfI = [10*cos(t1 + t2) 10*sin(t1 + t2); -10*cos(t1 + t2)-10*cos(t1) -10*sin(t1 + t2)-10*sin(t1)] / det
v = [-0.25; 0.5]
println(JfI * v)

[0.054364008549266075, -0.05129558466609972]


$$ \frac{d\theta_1}{dt} = 0.0544s^{-1}, \frac{d\theta_2}{dt} = -0.0513s^{-1} $$

#### 22. Modify the robot arm found in Fig. 3.6.1 where the base joint raises the arm by $d_1$ in the z direction. In addition, the joint mounts the arm along the rotation direction by $a_1$. When the base joint angle is zero, $\theta_1 = 0$, the two link arms indicated by $a_2$, $a_3$ lie in the x-z plane. The latter two joints would have an axis of rotation parallel to the y axis (again when $\theta_1 = 0$).

##### a. What is the DH parameter table?
<table style="font-size:15px;">
    <tr style="border-bottom:2px solid black;">
        <th style="text-align:center;">Link</th>
        <th style="text-align:center;">$\theta$</th>
        <th style="text-align:center;">$d$</th>
        <th style="text-align:center;">$a$</th>
        <th style="text-align:center;">$\alpha$</th>
    </tr>
    <tr >
        <td style="text-align:center;">1</td>
        <td style="text-align:center;">$\theta_1$</td>
        <td style="text-align:center;">$d_1$</td>
        <td style="text-align:center;">$a_1$</td>
        <td style="text-align:center;">$\pi / 2$</td>
    </tr>
    <tr>
        <td style="text-align:center;">2</td>
        <td style="text-align:center;">$\theta_2$</td>
        <td style="text-align:center;">$0$</td>
        <td style="text-align:center;">$a_2$</td>
        <td style="text-align:center;">$0$</td>
    </tr>
    <tr>
        <td style="text-align:center;">3</td>
        <td style="text-align:center;">$\theta_3$</td>
        <td style="text-align:center;">$0$</td>
        <td style="text-align:center;">$a_3$</td>
        <td style="text-align:center;">$0$</td>
    </tr>
</table>

##### b. Write out the DH transformation matrices.
$$
A_1 = \begin{pmatrix} cos(\theta_1) & -sin(\theta_1)cos(\frac{\pi}{2}) & sin(\theta_1)sin(\frac{\pi}{2}) & a_1cos(\theta_1) \\
sin(\theta_1) & cos(\theta_1)cos(\frac{\pi}{2}) & -cos(\theta_1)sin(\frac{\pi}{2}) & a_1sin(\theta_1) \\
0 & sin(\frac{\pi}{2}) & cos(\frac{\pi}{2}) & d_1 \\
0 & 0 & 0 & 1 \end{pmatrix} = \begin{pmatrix} cos(\theta_1) & 0 & sin(\theta_1) & a_1cos(\theta_1) \\
sin(\theta_1) & 0 & -cos(\theta_1) & a_1sin(\theta_1) \\
0 & 1 & 0 & d_1 \\
0 & 0 & 0 & 1 \end{pmatrix}
$$

$$
A_2 = \begin{pmatrix} cos(\theta_2) & -sin(\theta_2)cos(0) & sin(\theta_2)sin(0) & a_2cos(\theta_2) \\
sin(\theta_2) & cos(\theta_2)cos(0) & -cos(\theta_2)sin(0) & a_2sin(\theta_2) \\
0 & sin(0) & cos(0) & 0 \\
0 & 0 & 0 & 1 \end{pmatrix} = \begin{pmatrix} cos(\theta_2) & -sin(\theta_2) & 0 & a_2cos(\theta_2) \\
sin(\theta_2) & cos(\theta_2) & 0 & a_2sin(\theta_2) \\
0 & 0 & 1 & 0 \\
0 & 0 & 0 & 1 \end{pmatrix}
$$

$$
A_3 = \begin{pmatrix} cos(\theta_3) & -sin(\theta_3)cos(0) & sin(\theta_3)sin(0) & a_3cos(\theta_3) \\
sin(\theta_3) & cos(\theta_3)cos(0) & -cos(\theta_3)sin(0) & a_3sin(\theta_3) \\
0 & sin(0) & cos(0) & 0 \\
0 & 0 & 0 & 1 \end{pmatrix} = \begin{pmatrix} cos(\theta_3) & -sin(\theta_3) & 0 & a_3cos(\theta_3) \\
sin(\theta_3) & cos(\theta_3) & 0 & a_3sin(\theta_3) \\
0 & 0 & 1 & 0 \\
0 & 0 & 0 & 1 \end{pmatrix}
$$

##### c. Find the forward kinematics.
$$
T = \begin{pmatrix} cos(\theta_1) & 0 & sin(\theta_1) & a_1cos(\theta_1) \\
sin(\theta_1) & 0 & -cos(\theta_1) & a_1sin(\theta_1) \\
0 & 1 & 0 & d_1 \\
0 & 0 & 0 & 1 \end{pmatrix} \begin{pmatrix} cos(\theta_2) & -sin(\theta_2) & 0 & a_2cos(\theta_2) \\
sin(\theta_2) & cos(\theta_2) & 0 & a_2sin(\theta_2) \\
0 & 0 & 1 & 0 \\
0 & 0 & 0 & 1 \end{pmatrix} \begin{pmatrix} cos(\theta_3) & -sin(\theta_3) & 0 & a_3cos(\theta_3) \\
sin(\theta_3) & cos(\theta_3) & 0 & a_3sin(\theta_3) \\
0 & 0 & 1 & 0 \\
0 & 0 & 0 & 1 \end{pmatrix}
$$

$$
T = \begin{pmatrix} cos(\theta_1)cos(\theta_2) & -cos(\theta_1)sin(\theta_2) & sin(\theta_1) & a_2cos(\theta_1)cos(\theta_2) + a_1cos(\theta_1) \\
sin(\theta_1)cos(\theta_2) & -sin(\theta_1)sin(\theta_2) & -cos(\theta_1) & a_2sin(\theta_1)cos(\theta_2) + a_1sin(\theta_1) \\
sin(\theta_2) & cos(\theta_2) & 0 & a_2sin(\theta_2) + d_1 \\
0 & 0 & 0 & 1 \end{pmatrix} \begin{pmatrix} cos(\theta_3) & -sin(\theta_3) & 0 & a_3cos(\theta_3) \\
sin(\theta_3) & cos(\theta_3) & 0 & a_3sin(\theta_3) \\
0 & 0 & 1 & 0 \\
0 & 0 & 0 & 1 \end{pmatrix}
$$

$$
T = \begin{pmatrix} c_1c_2c_3 -c_1s_2s_3 & -c_1c_2s_3 - c_1s_2c_3 & s_1 & a_3c_1c_2c_3 -a_3c_1s_2s_3 + a_2c_1c_2 + a_1c_1 \\
s_1c_2c_3 -s_1s_2s_3 & -s_1c_2s_3 - s_1s_2c_3 & -c_1 & a_3s_1c_2c_3 -a_3s_1s_2s_3 + a_2s_1c_2 + a_1s_1 \\
s_2c_3 + c_2s_3 & -s_2s_3 + c_2c_3 & 0 & a_3s_2c_3 + a_3c_2s_3 + a_2s_2 + d_1 \\
0 & 0 & 0 & 1 \end{pmatrix}
$$

$$
T = \begin{pmatrix} c_1c_{23} & -c_1s_{23} & s_1 & (a_3c_{23} + a_2c_2 + a_1)c_1 \\
s_1c_{23} & -s_1s_{23} & -c_1 & (a_3c_{23} + a_2c_2 + a_1)s_1 \\
s_{23} & c_{23} & 0 & a_3s_{23} + a_2s_2 + d_1 \\
0 & 0 & 0 & 1 \end{pmatrix}
$$

#### 23. For the robotic arm found in Fig. 3.6.6, the distance $d_1$ is constant where $a_2$, $a_3$ can vary.

##### a. What is the DH parameter table?
<table style="font-size:15px;">
    <tr style="border-bottom:2px solid black;">
        <th style="text-align:center;">Link</th>
        <th style="text-align:center;">$\theta$</th>
        <th style="text-align:center;">$d$</th>
        <th style="text-align:center;">$a$</th>
        <th style="text-align:center;">$\alpha$</th>
    </tr>
    <tr >
        <td style="text-align:center;">2</td>
        <td style="text-align:center;">$\theta_1$</td>
        <td style="text-align:center;">$d_1$</td>
        <td style="text-align:center;">$a_2$</td>
        <td style="text-align:center;">$-\pi / 2$</td>
    </tr>
    <tr>
        <td style="text-align:center;">3</td>
        <td style="text-align:center;">$\pi / 2$</td>
        <td style="text-align:center;">$0$</td>
        <td style="text-align:center;">$a_3$</td>
        <td style="text-align:center;">$0$</td>
    </tr>
</table>

##### b. Write out the DH transformation matrices.
$$
A_2 = \begin{pmatrix} cos(\theta_1) & -sin(\theta_1)cos(\frac{-\pi}{2}) & sin(\theta_1)sin(\frac{-\pi}{2}) & a_2cos(\theta_1) \\
sin(\theta_1) & cos(\theta_1)cos(\frac{-\pi}{2}) & -cos(\theta_1)sin(\frac{-\pi}{2}) & a_2sin(\theta_1) \\
0 & sin(\frac{-\pi}{2}) & cos(\frac{-\pi}{2}) & d_1 \\
0 & 0 & 0 & 1 \end{pmatrix} = \begin{pmatrix} cos(\theta_1) & 0 & -sin(\theta_1) & a_2cos(\theta_1) \\
sin(\theta_1) & 0 & cos(\theta_1) & a_2sin(\theta_1) \\
0 & -1 & 0 & d_1 \\
0 & 0 & 0 & 1 \end{pmatrix}
$$

$$
A_3 = \begin{pmatrix} cos(\frac{\pi}{2}) & -sin(\frac{\pi}{2})cos(0) & sin(\frac{\pi}{2})sin(0) & a_3cos(\frac{\pi}{2}) \\
sin(\frac{\pi}{2}) & cos(\frac{\pi}{2})cos(0) & -cos(\frac{\pi}{2})sin(0) & a_3sin(\frac{\pi}{2}) \\
0 & sin(0) & cos(0) & 0 \\
0 & 0 & 0 & 1 \end{pmatrix} = \begin{pmatrix} 0 & -1 & 0 & 0 \\
1 & 0 & 0 & a_3 \\
0 & 0 & 1 & 0 \\
0 & 0 & 0 & 1 \end{pmatrix}
$$

##### c. Find the forward kinematics.
$$
T = \begin{pmatrix} cos(\theta_1) & 0 & -sin(\theta_1) & a_2cos(\theta_1) \\
sin(\theta_1) & 0 & cos(\theta_1) & a_2sin(\theta_1) \\
0 & -1 & 0 & d_1 \\
0 & 0 & 0 & 1 \end{pmatrix} \begin{pmatrix} 0 & -1 & 0 & 0 \\
1 & 0 & 0 & a_3 \\
0 & 0 & 1 & 0 \\
0 & 0 & 0 & 1 \end{pmatrix}
$$

$$
T = \begin{pmatrix} 0 & -cos(\theta_1) & -sin(\theta_1) & a_2cos(\theta_1) \\
0 & -sin(\theta_1) & cos(\theta_1) & a_2sin(\theta_1) \\
-1 & 0 & 0 & d_1 - a_3 \\
0 & 0 & 0 & 1 \end{pmatrix}
$$

#### 26. For the differential drive robot, let $r = 10$, $L = 15$, $\dot\phi_1 = 0.9$, $\dot\phi_2 = 1.2$.

##### a. What is the angular velocity of the robot?
$$ \dot\theta = \frac{10}{2 \cdot 15}(0.9 - 1.2) = -0.1s^{-1} $$

##### b. What is the velocity vector for the robot when $\theta = 45^{\circ}$?
$$ v = \frac{10}{2}(0.9 + 1.2) \begin{pmatrix} cos(45^{\circ}) \\ sin(45^{\circ}) \end{pmatrix} = \frac{21\sqrt{2}}{4} \begin{pmatrix} 1 \\ 1 \end{pmatrix} $$

#### 27. Let $r = 10$, $L = 15$. If you program the robot to drive straight and the robot traces out a circle of diameter 3 while traveling 1$s^{-1}$, what are the two wheel speeds?

$$ t = \frac{2\pi\frac{3}{2}}{1} = 3\pi $$

$$ \dot x_1 = \frac{2\pi\frac{3 + 15}{2}}{3\pi} = 6 $$

$$ \dot\phi_1 = \frac{6}{10} = \frac{3}{5} $$

$$ \dot\phi_2 = 2\frac{1}{10} - \frac{3}{5} = \frac{-2}{5} $$